Readme.md
# vimeo.js

Rewrite of `froogaloop.js` from vimeo.
Since it seems not maintained and no tests, I re-wrote it so it's more reliable.

## Installation

For [component(1)](http://github.com/component/component)
```
$ component install mnmly-vimeo
```

For simple script tag

```
<script src='/vimeo.js'></script>
<script type="text/javascript">
  console.log(Vimeo);
</script>
```

## Usage

```javascript
var Vimeo = require('vimeo');

// Start listening for message event from vimeo iframe
Vimeo.listen();

var vimeo = new Vimeo('vimeo-iframe-id');

// Instead of id of iframe, you can pass element as well
// vimeo = new Vimeo(iframeElement);

vimeo.on('ready', function(){
  console.log('ready to use api');

  vimeo.api('getDuration', function(d){
    console.log(d.value); // -> duration in seconds
  });
});
```


## API

  - [Vimeo()](#vimeo)
  - [Vimeo::on()](#vimeooneventnamestringfnfunction)
  - [Vimeo::api()](#vimeoapimethodstringvalfunctionstring)
  - [Vimeo::destroy()](#vimeodestroyremoveboolean)
  - [Vimeo.listen()](#vimeolisten)
  - [Vimeo.unlisten()](#vimeounlisten)

### Vimeo()

  Instantiate `Vimeo`

### Vimeo::on(eventName:String, fn:Function)

  Add event listeners and tells iFrame to emit the event.

### Vimeo::api(method:String, val:Function|String)
  
  API call.

### Vimeo::destroy(remove:Boolean)

  Destroys instance

### Vimeo::listen()

  Start listening for message event from iframe

### Vimeo.unlisten()

  Stop listening for message event from iframe

## API list

- *Getters*
     - `currentTime`
     - `getVideoHeight`
     - `getVideoWidth`
     - `getDuration`
     - `getColor`
     - `getVideoUrl`
     - `getVideoEmbedCode`
     - `getVolume`
     - `paused`

- *Setters*
     - `setLoop`
     - `setColor`
     - `setVolume`

- *Methods*
     - `play`
     - `pause`
     - `unload`
     - `seekTo`

## Events
  
  - `ready`
  - `play`
  - `pause`
  - `finish`
  - `seek`
  - `loadProgress`
  - `playProgress`

## TODO:

- examples


## License

  MIT


History.md
0.0.2 / 2014-02-16 
==================

Add random player id generation


0.0.1 / 2014-02-16 
==================

Initial commit for rewrite



test/tests.js
var chai = require('chai');
var should = chai.should();
var Debug = require('visionmedia-debug');

Debug.enable('*');

var Vimeo = require('vimeo');
Vimeo.listen();

describe('Vimeo', function(){

  this.timeout(5000);

  var demoSrc = '//player.vimeo.com/video/7100569?api=1&player_id=player_1';
  beforeEach(function(){
    this.iframe = document.createElement('iframe');
    this.iframe.id = 'test';
    this.iframe.src = demoSrc;
    document.body.appendChild(this.iframe);
    this.vimeo = new Vimeo('test');
  });

  afterEach(function(){
    document.body.removeChild(this.iframe);
    this.vimeo.destroy();
  });

  it('should exist', function(){
    should.exist(Vimeo);
  });

  it('should grab el when id is passed to constructor', function(){
    this.vimeo.should.have.a.property('el', this.iframe);
  });
  
  it('should grab a player domain', function(){
    this.vimeo.should.have.a.property('playerDomain');
    this.vimeo.playerDomain.should.equal('http://player.vimeo.com');
  });

  it('should trigger ready when ready', function(done){
    this.vimeo.on('ready', function(){
      done();
    });
  });

  describe('API', function(){

    beforeEach(function(next){
      this.vimeo.on('ready', function(){
        next();
      });
    });

    it('should have api/postMessage method', function(){
      this.vimeo.should.have.a.property('api');
      this.vimeo.should.have.a.property('postMessage');
    });

    describe('Getters', function(){
      
      it('should get currentTime', function(done){
        this.vimeo.api('getCurrentTime', function(e){
          e.value.should.be.number;
          done();
        });
      });

      it('should get video width', function(done){
        this.vimeo.api('getVideoHeight', function(e){
          e.value.should.be.number;
          done();
        });
      });

      it('should get video height', function(done){
        this.vimeo.api('getVideoWidth', function(e){
          e.value.should.be.number;
          done();
        });
      });

      it('should get duration', function(done){
        this.vimeo.api('getDuration', function(e){
          e.value.should.be.number;
          done();
        });
      });
      
      it('should get color', function(done){
        this.vimeo.api('getColor', function(e){
          var color = '#' + e.value;
          color.should.match(/^#[0-9A-F]{6}$/i);
          done();
        });
      });
      
      it('should get url', function(done){
        this.vimeo.api('getVideoUrl', function(e){
          e.value.should.match(/^http/);
          done();
        });
      });
      
      it('should get embed code', function(done){
        this.vimeo.api('getVideoEmbedCode', function(e){
          e.value.should.match(/iframe/g);
          done();
        });
      });
      
      it('should get paused or not', function(done){
        this.vimeo.api('paused', function(e){
          e.value.should.be.ok;
          done();
        });
      });
      
      it('should get volume', function(done){
        this.vimeo.api('getVolume', function(e){
          e.value.should.not.be.number;
          done();
        });
      });


    });
    
    describe('Setters', function(){
      it('should set volume', function(done){
        var self = this;
        var volume = 0.1;
        this.vimeo.api('setVolume', volume);
        setTimeout(function(){
          self.vimeo.api('getVolume', function(e){
            e.value.should.equal(volume);
            done();
          });
        }, 1000);
      });

      it('should set loop', function(done){
        var self = this;
        this.vimeo.api('setLoop', 10);
        done();
      });
      
      it('should set color', function(done){
        var self = this;
        var hex = '#000000'.replace(/#/, '');
        this.vimeo.api('setColor', hex);

        setTimeout(function(){
          self.vimeo.api('getColor', function(e){
            e.value.should.equal(hex);
            done();
          });
        }, 500);
      });
    });

    describe('Methods', function(){

      it('should play', function(done){

        this.vimeo.on('play', function(){
          true.should.be.okay;
          done();
        });
        this.vimeo.api('play');
      });
      
      it('should pause', function(done){
        var self = this;
        this.vimeo.on('pause', function(){
          true.should.be.okay;
          done();
        });

        this.vimeo.api('play');
        setTimeout(function(){
          self.vimeo.api('pause');
        }, 1000);
      });
      
      it('should seek', function(done){
        var sec = 10;
        this.vimeo.on('seek', function(e){
          e.data.seconds.should.equal(sec);
          done();
        });
        this.vimeo.api('seekTo', sec);
      });
      
      it('should unload', function(done){
        var self = this;
        this.vimeo.api('play');
        setTimeout(function(){
          self.vimeo.api('unload');
          done();
        }, 3000);
      });

      it('should notify when video finishes', function(done){
          var self = this;
          this.timeout(10000);
          this.vimeo.api('getDuration', function(e){
            var duration = e.value;
            self.vimeo.api('seekTo', duration - 2);
            self.vimeo.on('finish', function(){
              done();
            });
          });
      });
    });
  });
});


index.js
/**
 * Module dependencies
 */

var events = require('event');
var Emitter = require('emitter');
var url = require('url');
var qs = require('querystring');
var debug = require('debug')('vimeo');
var find = require('find');
var type = require('type');

/**
 * Expose `Vimeo`
 */

module.exports = Vimeo;

/**
 * Instantiate `Vimeo`
 *
 * @param {Element/String} iframe
 */

function Vimeo(iframe) {

  if(typeof iframe === 'string') iframe = document.getElementById(iframe);

  this.el = iframe;
  this.ready = false;
  this.src = this.el.getAttribute('src');
  var u = url.parse(this.src);
  this.playerDomain = u.protocol + '//' + u.host;
  this.id = qs.parse(u.query.replace(/&amp;/g, '&')).player_id;

  if(!this.id) {
    this.id = Vimeo.randomPlayerId();
    this.src += '&amp;player_id=' + this.id;
    this.el.setAttribute('src', this.src);
  }

  Vimeo.Collection.push(this);

}

/**
 * Collection of instance
 * @api private
 */

Vimeo.Collection = [];

/**
 * Installs Emitter
 */

Emitter(Vimeo.prototype);


/**
 * Add event listeners and tells iFrame to emit the event.
 *
 * @param {String} eventName
 * @param {Function} fn
 * @api public
 */

Vimeo.prototype.on = function(eventName, fn){

  var rtn = this._on.apply(this, arguments);

  if(eventName === 'ready'){
    if(this.ready) fn.call(this, { event: 'ready', 'player_id': this.id });
  } else {
    this.postMessage('addEventListener', eventName);
  }

  return rtn;
};


/**
 * Manages events, Inherits from `Emitter`'s `on`
 *
 * @api private
 */

Vimeo.prototype._on = function(){
  return Emitter.prototype.on.apply(this, arguments);
};

/**
 * API
 *
 * Usage: 
 * ```
 *  // Getters
 *  vimeo.api('currentTime', function(d){
 *    console.log(d.value);
 *  });
 *
 *  // Setters
 *  vimeo.api('setLoop', 10);
 *
 *  // Methods
 *  vimeo.api('seekTo', 10);
 *
 * ```
 *  - *Getters*
 *      - `currentTime`
 *      - `getVideoHeight`
 *      - `getVideoWidth`
 *      - `getDuration`
 *      - `getColor`
 *      - `getVideoUrl`
 *      - `getVideoEmbedCode`
 *      - `getVolume`
 *      - `paused`
 *
 *  - *Setters*
 *      - `setLoop`
 *      - `setColor`
 *      - `setVolume`
 *
 *  - *Methods*
 *      - `play`
 *      - `pause`
 *      - `unload`
 *      - `seekTo`
 *
 * @param {String} method
 * @param {Function/String} val
 * @api public
 */

Vimeo.prototype.api = function(method, val){
  if('function' === type(val)){
    this._on(method, val); 
    val = null;
  }
  this.postMessage(method, val);
};


/**
 * Post message to iframe
 *
 * @param {String} method
 * @param {String} val
 * @api private
 */

Vimeo.prototype.postMessage = function(method, val){

  if(!this.el.contentWindow.postMessage) return false;
  
  var u = url.parse(this.src);
  var _url = u.protocol + '//' + u.hostname + u.pathname;
  var payload = { method: method, value: val };
  var data = JSON.stringify(payload);

  debug('postMessage payload: %s', data);
  debug('postMessage target url: %s', _url);
  
  this.el.contentWindow.postMessage(data, _url);

};


/**
 * Destroys instance
 *
 * @param {Boolean} remove
 * @api public
 */

Vimeo.prototype.destroy = function(remove){
  this.off();
  var idx = Vimeo.Collection.indexOf(this);
  if(idx > -1) Vimeo.Collection.splice(idx, 1);
  if(remove) this.el.parentElement.removeChild(this.el);
};


/**
 * Callback for `message` event from vimeo iframe
 * @param {Event} e
 *
 * @api private
 */

Vimeo.onMessageRecieved = function(e){

  debug('recieveMessage payload: %s', e.data);

  var data = JSON.parse(e.data);
  var eventName = data.event || data.method;
  
  if(data.ready) this.ready = true;
  if(!data.player_id) return;

  var instance = find(Vimeo.Collection, function(d){
    return d.id === data.player_id;
  });
  
  if(e.origin !== instance.playerDomain) return false;
  instance.emit(eventName, data);

};

/**
 * Start listening for message event from iframe
 *
 * @api public
 */

Vimeo.listen = function(){
  events.bind(window, 'message', Vimeo.onMessageRecieved);
};

/**
 * Stop listening for message event from iframe
 *
 * @api public
 */

Vimeo.unlisten = function(){
  events.unbind(window, 'message', Vimeo.onMessageRecieved);
};

Vimeo.randomPlayerId = function(){
  return 'player-' + Math.random().toString(36).substring(7);
};


vimeo.js
;(function(){

/**
 * Require the given path.
 *
 * @param {String} path
 * @return {Object} exports
 * @api public
 */

function require(path, parent, orig) {
  var resolved = require.resolve(path);

  // lookup failed
  if (null == resolved) {
    orig = orig || path;
    parent = parent || 'root';
    var err = new Error('Failed to require "' + orig + '" from "' + parent + '"');
    err.path = orig;
    err.parent = parent;
    err.require = true;
    throw err;
  }

  var module = require.modules[resolved];

  // perform real require()
  // by invoking the module's
  // registered function
  if (!module._resolving && !module.exports) {
    var mod = {};
    mod.exports = {};
    mod.client = mod.component = true;
    module._resolving = true;
    module.call(this, mod.exports, require.relative(resolved), mod);
    delete module._resolving;
    module.exports = mod.exports;
  }

  return module.exports;
}

/**
 * Registered modules.
 */

require.modules = {};

/**
 * Registered aliases.
 */

require.aliases = {};

/**
 * Resolve `path`.
 *
 * Lookup:
 *
 *   - PATH/index.js
 *   - PATH.js
 *   - PATH
 *
 * @param {String} path
 * @return {String} path or null
 * @api private
 */

require.resolve = function(path) {
  if (path.charAt(0) === '/') path = path.slice(1);

  var paths = [
    path,
    path + '.js',
    path + '.json',
    path + '/index.js',
    path + '/index.json'
  ];

  for (var i = 0; i < paths.length; i++) {
    var path = paths[i];
    if (require.modules.hasOwnProperty(path)) return path;
    if (require.aliases.hasOwnProperty(path)) return require.aliases[path];
  }
};

/**
 * Normalize `path` relative to the current path.
 *
 * @param {String} curr
 * @param {String} path
 * @return {String}
 * @api private
 */

require.normalize = function(curr, path) {
  var segs = [];

  if ('.' != path.charAt(0)) return path;

  curr = curr.split('/');
  path = path.split('/');

  for (var i = 0; i < path.length; ++i) {
    if ('..' == path[i]) {
      curr.pop();
    } else if ('.' != path[i] && '' != path[i]) {
      segs.push(path[i]);
    }
  }

  return curr.concat(segs).join('/');
};

/**
 * Register module at `path` with callback `definition`.
 *
 * @param {String} path
 * @param {Function} definition
 * @api private
 */

require.register = function(path, definition) {
  require.modules[path] = definition;
};

/**
 * Alias a module definition.
 *
 * @param {String} from
 * @param {String} to
 * @api private
 */

require.alias = function(from, to) {
  if (!require.modules.hasOwnProperty(from)) {
    throw new Error('Failed to alias "' + from + '", it does not exist');
  }
  require.aliases[to] = from;
};

/**
 * Return a require function relative to the `parent` path.
 *
 * @param {String} parent
 * @return {Function}
 * @api private
 */

require.relative = function(parent) {
  var p = require.normalize(parent, '..');

  /**
   * lastIndexOf helper.
   */

  function lastIndexOf(arr, obj) {
    var i = arr.length;
    while (i--) {
      if (arr[i] === obj) return i;
    }
    return -1;
  }

  /**
   * The relative require() itself.
   */

  function localRequire(path) {
    var resolved = localRequire.resolve(path);
    return require(resolved, parent, path);
  }

  /**
   * Resolve relative to the parent.
   */

  localRequire.resolve = function(path) {
    var c = path.charAt(0);
    if ('/' == c) return path.slice(1);
    if ('.' == c) return require.normalize(p, path);

    // resolve deps by returning
    // the dep in the nearest "deps"
    // directory
    var segs = parent.split('/');
    var i = lastIndexOf(segs, 'deps') + 1;
    if (!i) i = 0;
    path = segs.slice(0, i + 1).join('/') + '/deps/' + path;
    return path;
  };

  /**
   * Check if module is defined at `path`.
   */

  localRequire.exists = function(path) {
    return require.modules.hasOwnProperty(localRequire.resolve(path));
  };

  return localRequire;
};
require.register("component-event/index.js", function(exports, require, module){
var bind = window.addEventListener ? 'addEventListener' : 'attachEvent',
    unbind = window.removeEventListener ? 'removeEventListener' : 'detachEvent',
    prefix = bind !== 'addEventListener' ? 'on' : '';

/**
 * Bind `el` event `type` to `fn`.
 *
 * @param {Element} el
 * @param {String} type
 * @param {Function} fn
 * @param {Boolean} capture
 * @return {Function}
 * @api public
 */

exports.bind = function(el, type, fn, capture){
  el[bind](prefix + type, fn, capture || false);
  return fn;
};

/**
 * Unbind `el` event `type`'s callback `fn`.
 *
 * @param {Element} el
 * @param {String} type
 * @param {Function} fn
 * @param {Boolean} capture
 * @return {Function}
 * @api public
 */

exports.unbind = function(el, type, fn, capture){
  el[unbind](prefix + type, fn, capture || false);
  return fn;
};
});
require.register("component-emitter/index.js", function(exports, require, module){

/**
 * Expose `Emitter`.
 */

module.exports = Emitter;

/**
 * Initialize a new `Emitter`.
 *
 * @api public
 */

function Emitter(obj) {
  if (obj) return mixin(obj);
};

/**
 * Mixin the emitter properties.
 *
 * @param {Object} obj
 * @return {Object}
 * @api private
 */

function mixin(obj) {
  for (var key in Emitter.prototype) {
    obj[key] = Emitter.prototype[key];
  }
  return obj;
}

/**
 * Listen on the given `event` with `fn`.
 *
 * @param {String} event
 * @param {Function} fn
 * @return {Emitter}
 * @api public
 */

Emitter.prototype.on =
Emitter.prototype.addEventListener = function(event, fn){
  this._callbacks = this._callbacks || {};
  (this._callbacks[event] = this._callbacks[event] || [])
    .push(fn);
  return this;
};

/**
 * Adds an `event` listener that will be invoked a single
 * time then automatically removed.
 *
 * @param {String} event
 * @param {Function} fn
 * @return {Emitter}
 * @api public
 */

Emitter.prototype.once = function(event, fn){
  var self = this;
  this._callbacks = this._callbacks || {};

  function on() {
    self.off(event, on);
    fn.apply(this, arguments);
  }

  on.fn = fn;
  this.on(event, on);
  return this;
};

/**
 * Remove the given callback for `event` or all
 * registered callbacks.
 *
 * @param {String} event
 * @param {Function} fn
 * @return {Emitter}
 * @api public
 */

Emitter.prototype.off =
Emitter.prototype.removeListener =
Emitter.prototype.removeAllListeners =
Emitter.prototype.removeEventListener = function(event, fn){
  this._callbacks = this._callbacks || {};

  // all
  if (0 == arguments.length) {
    this._callbacks = {};
    return this;
  }

  // specific event
  var callbacks = this._callbacks[event];
  if (!callbacks) return this;

  // remove all handlers
  if (1 == arguments.length) {
    delete this._callbacks[event];
    return this;
  }

  // remove specific handler
  var cb;
  for (var i = 0; i < callbacks.length; i++) {
    cb = callbacks[i];
    if (cb === fn || cb.fn === fn) {
      callbacks.splice(i, 1);
      break;
    }
  }
  return this;
};

/**
 * Emit `event` with the given args.
 *
 * @param {String} event
 * @param {Mixed} ...
 * @return {Emitter}
 */

Emitter.prototype.emit = function(event){
  this._callbacks = this._callbacks || {};
  var args = [].slice.call(arguments, 1)
    , callbacks = this._callbacks[event];

  if (callbacks) {
    callbacks = callbacks.slice(0);
    for (var i = 0, len = callbacks.length; i < len; ++i) {
      callbacks[i].apply(this, args);
    }
  }

  return this;
};

/**
 * Return array of callbacks for `event`.
 *
 * @param {String} event
 * @return {Array}
 * @api public
 */

Emitter.prototype.listeners = function(event){
  this._callbacks = this._callbacks || {};
  return this._callbacks[event] || [];
};

/**
 * Check if this emitter has `event` handlers.
 *
 * @param {String} event
 * @return {Boolean}
 * @api public
 */

Emitter.prototype.hasListeners = function(event){
  return !! this.listeners(event).length;
};

});
require.register("component-url/index.js", function(exports, require, module){

/**
 * Parse the given `url`.
 *
 * @param {String} str
 * @return {Object}
 * @api public
 */

exports.parse = function(url){
  var a = document.createElement('a');
  a.href = url;
  return {
    href: a.href,
    host: a.host || location.host,
    port: ('0' === a.port || '' === a.port) ? port(a.protocol) : a.port,
    hash: a.hash,
    hostname: a.hostname || location.hostname,
    pathname: a.pathname.charAt(0) != '/' ? '/' + a.pathname : a.pathname,
    protocol: !a.protocol || ':' == a.protocol ? location.protocol : a.protocol,
    search: a.search,
    query: a.search.slice(1)
  };
};

/**
 * Check if `url` is absolute.
 *
 * @param {String} url
 * @return {Boolean}
 * @api public
 */

exports.isAbsolute = function(url){
  return 0 == url.indexOf('//') || !!~url.indexOf('://');
};

/**
 * Check if `url` is relative.
 *
 * @param {String} url
 * @return {Boolean}
 * @api public
 */

exports.isRelative = function(url){
  return !exports.isAbsolute(url);
};

/**
 * Check if `url` is cross domain.
 *
 * @param {String} url
 * @return {Boolean}
 * @api public
 */

exports.isCrossDomain = function(url){
  url = exports.parse(url);
  return url.hostname !== location.hostname
    || url.port !== location.port
    || url.protocol !== location.protocol;
};

/**
 * Return default port for `protocol`.
 *
 * @param  {String} protocol
 * @return {String}
 * @api private
 */
function port (protocol){
  switch (protocol) {
    case 'http:':
      return 80;
    case 'https:':
      return 443;
    default:
      return location.port;
  }
}

});
require.register("component-trim/index.js", function(exports, require, module){

exports = module.exports = trim;

function trim(str){
  if (str.trim) return str.trim();
  return str.replace(/^\s*|\s*$/g, '');
}

exports.left = function(str){
  if (str.trimLeft) return str.trimLeft();
  return str.replace(/^\s*/, '');
};

exports.right = function(str){
  if (str.trimRight) return str.trimRight();
  return str.replace(/\s*$/, '');
};

});
require.register("component-querystring/index.js", function(exports, require, module){

/**
 * Module dependencies.
 */

var trim = require('trim');

/**
 * Parse the given query `str`.
 *
 * @param {String} str
 * @return {Object}
 * @api public
 */

exports.parse = function(str){
  if ('string' != typeof str) return {};

  str = trim(str);
  if ('' == str) return {};
  if ('?' == str.charAt(0)) str = str.slice(1);

  var obj = {};
  var pairs = str.split('&');
  for (var i = 0; i < pairs.length; i++) {
    var parts = pairs[i].split('=');
    obj[parts[0]] = null == parts[1]
      ? ''
      : decodeURIComponent(parts[1]);
  }

  return obj;
};

/**
 * Stringify the given `obj`.
 *
 * @param {Object} obj
 * @return {String}
 * @api public
 */

exports.stringify = function(obj){
  if (!obj) return '';
  var pairs = [];
  for (var key in obj) {
    pairs.push(encodeURIComponent(key) + '=' + encodeURIComponent(obj[key]));
  }
  return pairs.join('&');
};

});
require.register("component-props/index.js", function(exports, require, module){
/**
 * Global Names
 */

var globals = /\b(Array|Date|Object|Math|JSON)\b/g;

/**
 * Return immediate identifiers parsed from `str`.
 *
 * @param {String} str
 * @param {String|Function} map function or prefix
 * @return {Array}
 * @api public
 */

module.exports = function(str, fn){
  var p = unique(props(str));
  if (fn && 'string' == typeof fn) fn = prefixed(fn);
  if (fn) return map(str, p, fn);
  return p;
};

/**
 * Return immediate identifiers in `str`.
 *
 * @param {String} str
 * @return {Array}
 * @api private
 */

function props(str) {
  return str
    .replace(/\.\w+|\w+ *\(|"[^"]*"|'[^']*'|\/([^/]+)\//g, '')
    .replace(globals, '')
    .match(/[a-zA-Z_]\w*/g)
    || [];
}

/**
 * Return `str` with `props` mapped with `fn`.
 *
 * @param {String} str
 * @param {Array} props
 * @param {Function} fn
 * @return {String}
 * @api private
 */

function map(str, props, fn) {
  var re = /\.\w+|\w+ *\(|"[^"]*"|'[^']*'|\/([^/]+)\/|[a-zA-Z_]\w*/g;
  return str.replace(re, function(_){
    if ('(' == _[_.length - 1]) return fn(_);
    if (!~props.indexOf(_)) return _;
    return fn(_);
  });
}

/**
 * Return unique array.
 *
 * @param {Array} arr
 * @return {Array}
 * @api private
 */

function unique(arr) {
  var ret = [];

  for (var i = 0; i < arr.length; i++) {
    if (~ret.indexOf(arr[i])) continue;
    ret.push(arr[i]);
  }

  return ret;
}

/**
 * Map with prefix `str`.
 */

function prefixed(str) {
  return function(_){
    return str + _;
  };
}

});
require.register("component-to-function/index.js", function(exports, require, module){
/**
 * Module Dependencies
 */

try {
  var expr = require('props');
} catch(e) {
  var expr = require('props-component');
}

/**
 * Expose `toFunction()`.
 */

module.exports = toFunction;

/**
 * Convert `obj` to a `Function`.
 *
 * @param {Mixed} obj
 * @return {Function}
 * @api private
 */

function toFunction(obj) {
  switch ({}.toString.call(obj)) {
    case '[object Object]':
      return objectToFunction(obj);
    case '[object Function]':
      return obj;
    case '[object String]':
      return stringToFunction(obj);
    case '[object RegExp]':
      return regexpToFunction(obj);
    default:
      return defaultToFunction(obj);
  }
}

/**
 * Default to strict equality.
 *
 * @param {Mixed} val
 * @return {Function}
 * @api private
 */

function defaultToFunction(val) {
  return function(obj){
    return val === obj;
  }
}

/**
 * Convert `re` to a function.
 *
 * @param {RegExp} re
 * @return {Function}
 * @api private
 */

function regexpToFunction(re) {
  return function(obj){
    return re.test(obj);
  }
}

/**
 * Convert property `str` to a function.
 *
 * @param {String} str
 * @return {Function}
 * @api private
 */

function stringToFunction(str) {
  // immediate such as "> 20"
  if (/^ *\W+/.test(str)) return new Function('_', 'return _ ' + str);

  // properties such as "name.first" or "age > 18" or "age > 18 && age < 36"
  return new Function('_', 'return ' + get(str));
}

/**
 * Convert `object` to a function.
 *
 * @param {Object} object
 * @return {Function}
 * @api private
 */

function objectToFunction(obj) {
  var match = {}
  for (var key in obj) {
    match[key] = typeof obj[key] === 'string'
      ? defaultToFunction(obj[key])
      : toFunction(obj[key])
  }
  return function(val){
    if (typeof val !== 'object') return false;
    for (var key in match) {
      if (!(key in val)) return false;
      if (!match[key](val[key])) return false;
    }
    return true;
  }
}

/**
 * Built the getter function. Supports getter style functions
 *
 * @param {String} str
 * @return {String}
 * @api private
 */

function get(str) {
  var props = expr(str);
  if (!props.length) return '_.' + str;

  var val;
  for(var i = 0, prop; prop = props[i]; i++) {
    val = '_.' + prop;
    val = "('function' == typeof " + val + " ? " + val + "() : " + val + ")";
    str = str.replace(new RegExp(prop, 'g'), val);
  }

  return str;
}

});
require.register("component-find/index.js", function(exports, require, module){

/**
 * Module dependencies.
 */

var toFunction = require('to-function');

/**
 * Find the first value in `arr` with when `fn(val, i)` is truthy.
 *
 * @param {Array} arr
 * @param {Function} fn
 * @return {Array}
 * @api public
 */

module.exports = function(arr, fn){
  // callback
  if ('function' != typeof fn) {
    if (Object(fn) === fn) fn = objectToFunction(fn);
    else fn = toFunction(fn);
  }

  // filter
  for (var i = 0, len = arr.length; i < len; ++i) {
    if (fn(arr[i], i)) return arr[i];
  }
};

/**
 * Convert `obj` into a match function.
 *
 * @param {Object} obj
 * @return {Function}
 * @api private
 */

function objectToFunction(obj) {
  return function(o){
    for (var key in obj) {
      if (o[key] != obj[key]) return false;
    }
    return true;
  }
}
});
require.register("component-type/index.js", function(exports, require, module){
/**
 * toString ref.
 */

var toString = Object.prototype.toString;

/**
 * Return the type of `val`.
 *
 * @param {Mixed} val
 * @return {String}
 * @api public
 */

module.exports = function(val){
  switch (toString.call(val)) {
    case '[object Date]': return 'date';
    case '[object RegExp]': return 'regexp';
    case '[object Arguments]': return 'arguments';
    case '[object Array]': return 'array';
    case '[object Error]': return 'error';
  }

  if (val === null) return 'null';
  if (val === undefined) return 'undefined';
  if (val !== val) return 'nan';
  if (val && val.nodeType === 1) return 'element';

  return typeof val.valueOf();
};

});
require.register("visionmedia-debug/debug.js", function(exports, require, module){

/**
 * Expose `debug()` as the module.
 */

module.exports = debug;

/**
 * Create a debugger with the given `name`.
 *
 * @param {String} name
 * @return {Type}
 * @api public
 */

function debug(name) {
  if (!debug.enabled(name)) return function(){};

  return function(fmt){
    fmt = coerce(fmt);

    var curr = new Date;
    var ms = curr - (debug[name] || curr);
    debug[name] = curr;

    fmt = name
      + ' '
      + fmt
      + ' +' + debug.humanize(ms);

    // This hackery is required for IE8
    // where `console.log` doesn't have 'apply'
    window.console
      && console.log
      && Function.prototype.apply.call(console.log, console, arguments);
  }
}

/**
 * The currently active debug mode names.
 */

debug.names = [];
debug.skips = [];

/**
 * Enables a debug mode by name. This can include modes
 * separated by a colon and wildcards.
 *
 * @param {String} name
 * @api public
 */

debug.enable = function(name) {
  try {
    localStorage.debug = name;
  } catch(e){}

  var split = (name || '').split(/[\s,]+/)
    , len = split.length;

  for (var i = 0; i < len; i++) {
    name = split[i].replace('*', '.*?');
    if (name[0] === '-') {
      debug.skips.push(new RegExp('^' + name.substr(1) + '$'));
    }
    else {
      debug.names.push(new RegExp('^' + name + '$'));
    }
  }
};

/**
 * Disable debug output.
 *
 * @api public
 */

debug.disable = function(){
  debug.enable('');
};

/**
 * Humanize the given `ms`.
 *
 * @param {Number} m
 * @return {String}
 * @api private
 */

debug.humanize = function(ms) {
  var sec = 1000
    , min = 60 * 1000
    , hour = 60 * min;

  if (ms >= hour) return (ms / hour).toFixed(1) + 'h';
  if (ms >= min) return (ms / min).toFixed(1) + 'm';
  if (ms >= sec) return (ms / sec | 0) + 's';
  return ms + 'ms';
};

/**
 * Returns true if the given mode name is enabled, false otherwise.
 *
 * @param {String} name
 * @return {Boolean}
 * @api public
 */

debug.enabled = function(name) {
  for (var i = 0, len = debug.skips.length; i < len; i++) {
    if (debug.skips[i].test(name)) {
      return false;
    }
  }
  for (var i = 0, len = debug.names.length; i < len; i++) {
    if (debug.names[i].test(name)) {
      return true;
    }
  }
  return false;
};

/**
 * Coerce `val`.
 */

function coerce(val) {
  if (val instanceof Error) return val.stack || val.message;
  return val;
}

// persist

try {
  if (window.localStorage) debug.enable(localStorage.debug);
} catch(e){}

});
require.register("froogaloop/index.js", function(exports, require, module){
/**
 * Module dependencies
 */

var events = require('event');
var Emitter = require('emitter');
var url = require('url');
var qs = require('querystring');
var debug = require('debug')('vimeo');
var find = require('find');
var type = require('type');

/**
 * Expose `Froogaloop`
 */

module.exports = Froogaloop;

function Froogaloop(iframe) {

  if(typeof iframe === 'string') iframe = document.getElementById(iframe);

  this.el = iframe;
  this.ready = false;
  this.src = this.el.getAttribute('src');
  var u = url.parse(this.src);
  this.playerDomain = u.protocol + '//' + u.host;
  this.id = qs.parse(u.query.replace(/&amp;/g, '&')).player_id;

  if(!this.id) {
    throw new Error('player_id is required for player src');
  }

  Froogaloop.Collection.push(this);

}

Froogaloop.Collection = [];

/**
 * Emitter
 */

Emitter(Froogaloop.prototype);

Froogaloop.prototype.on = function(eventName, fn){

  var rtn = this._on.apply(this, arguments);

  if(eventName === 'ready'){
    if(this.ready) fn.call(this, { event: 'ready', 'player_id': this.id });
  } else {
    this.postMessage('addEventListener', eventName);
  }

  return rtn;
};


/**
 * Actual event handling
 */

Froogaloop.prototype._on = function(){
  return Emitter.prototype.on.apply(this, arguments);
};

/**
 * Higher level api
 */

Froogaloop.prototype.api = function(method, val){
  if('function' === type(val)){
    this._on(method, val); 
    val = null;
  }
  this.postMessage(method, val);
};

Froogaloop.prototype.postMessage = function(method, params){

  if(!this.el.contentWindow.postMessage) return false;
  
  var u = url.parse(this.src);
  var _url = u.protocol + '//' + u.hostname + u.pathname;
  var payload = { method: method, value: params };
  var data = JSON.stringify(payload);

  debug('postMessage payload: %s', data);
  debug('postMessage target url: %s', _url);
  
  this.el.contentWindow.postMessage(data, _url);

};

Froogaloop.prototype.destroy = function(){
  this.off();
  var idx = Froogaloop.Collection.indexOf(this);
  if(idx > -1) Froogaloop.Collection.splice(idx, 1);
};

Froogaloop.onMessageRecieved = function(e){

  debug('recieveMessage payload: %s', e.data);

  var data = JSON.parse(e.data);
  var eventName = data.event || data.method;
  
  if(data.ready) this.ready = true;
  if(!data.player_id) return;

  var instance = find(Froogaloop.Collection, function(d){
    return d.id === data.player_id;
  });
  
  if(e.origin !== instance.playerDomain) return false;
  instance.emit(eventName, data);

};


Froogaloop.listen = function(){
  events.bind(window, 'message', Froogaloop.onMessageRecieved);
};

Froogaloop.unlisten = function(){
  events.unbind(window, 'message', Froogaloop.onMessageRecieved);
};


});














require.alias("component-event/index.js", "froogaloop/deps/event/index.js");
require.alias("component-event/index.js", "event/index.js");

require.alias("component-emitter/index.js", "froogaloop/deps/emitter/index.js");
require.alias("component-emitter/index.js", "emitter/index.js");

require.alias("component-url/index.js", "froogaloop/deps/url/index.js");
require.alias("component-url/index.js", "url/index.js");

require.alias("component-querystring/index.js", "froogaloop/deps/querystring/index.js");
require.alias("component-querystring/index.js", "querystring/index.js");
require.alias("component-trim/index.js", "component-querystring/deps/trim/index.js");

require.alias("component-find/index.js", "froogaloop/deps/find/index.js");
require.alias("component-find/index.js", "find/index.js");
require.alias("component-to-function/index.js", "component-find/deps/to-function/index.js");
require.alias("component-props/index.js", "component-to-function/deps/props/index.js");

require.alias("component-type/index.js", "froogaloop/deps/type/index.js");
require.alias("component-type/index.js", "type/index.js");

require.alias("visionmedia-debug/debug.js", "froogaloop/deps/debug/debug.js");
require.alias("visionmedia-debug/debug.js", "froogaloop/deps/debug/index.js");
require.alias("visionmedia-debug/debug.js", "debug/index.js");
require.alias("visionmedia-debug/debug.js", "visionmedia-debug/index.js");if (typeof exports == "object") {
  module.exports = require("froogaloop");
} else if (typeof define == "function" && define.amd) {
  define(function(){ return require("froogaloop"); });
} else {
  this["Vimeo"] = require("froogaloop");
}})();